<!-- about.html-->

{% extends "base.html" %}
{% block title %}About{% endblock %}
{% block body %}

<link rel= "stylesheet" type= "text/css" href= "{{ url_for('static', filename='css/about.css') }}">
<script src="{{ url_for('static', filename='js/about.js') }}"></script>

<div class="modal-body">
  <h2>Table of Contents:</h2>
  <ol>
    <li><a href="#section1">QBiC-Pred input</a></li>
    <li><a href="#section2">TF binding models</a></li>
    <li><a href="#section3">QBiC-Pred output</a></li>
  </ol>

  <a name="section1"><h3>1. QBiC-Pred input</h3></a>
  <p>The main input file for QBiC-Pred must be a text file containing variants/
    mutations in one of the following formats:</p>
  <ul>
    <li>
      <p>
        VCF format (<a href="/download/vcf-example.vcf">download sample file</a>) <br />
        Each line has at least 5 tab-separated fields. <br />
        Column 1: chromosome name (hg19 or hg38) <br />
        Column 2: position (hg19 or hg38); must be a positive integer within chromosome length <br />
        Column 3: variant name; this field can be left empty <br />
        Column 4: reference allele; must be a single upper or lower case nucleotide letter <br />
        Column 5: alternative allele; must be a single upper or lower case nucleotide letter <br />
         (future QBiC releases will also allow short indels and multiple base substitutions) <br />
      </p>
      <div class="table-responsive">
        <table class="table table-borderless example-table">
          <tbody>
            <tr>
              <td>chr1</td>
              <td>48473010</td>
              <td>.</td>
              <td>A</td>
              <td>G</td>
            </tr>
            <tr>
              <td>chr2</td>
              <td>97536853</td>
              <td>.</td>
              <td>C</td>
              <td>T</td>
            </tr>
            <tr>
              <td>chr3</td>
              <td>181842690</td>
              <td>.</td>
              <td>G</td>
              <td>A</td>
            </tr>
            <tr>
              <td>chr5</td>
              <td>54469459</td>
              <td>.</td>
              <td>C</td>
              <td>G</td>
            </tr>
          </tbody>
        </table>
      </div>
    </li><br />
    <li><p>
      ICGC simple somatic mutation file  (<a href="/download/icgc-example.tsv">download sample file</a>) <br />
      Please see the ICGC guidelines  for details about this tab-separated file format. The columns used by QBiC-Pred are: chromosome, chromosome_start, chromosome_end (which must be the same as chromosome_start for single nucleotide variants), mutation_type (QBiC-Pred will only use variants with this field set to “single base substitution”), mutated_from_allele, and mutated_to_allele.
    </p>
    <div class="table-responsive">
      <table class="table table-borderless example-table">
        <thead>
          <tr>
            <th>...</th>
            <th>chromosome</th>
            <th>chromosome_start</th>
            <th>chromosome_end</th>
            <th>...</th>
            <th>mutation_type</th>
            <th>...</th>
            <th>mutated_from_allele</th>
            <th>mutated_to_allele</th>
          </tr>
        </thead>
        <tr>
          <td>...</td>
          <td>6</td>
          <td>65152140</td>
          <td>65152140</td>
          <td>...</td>
          <td>single base substitution</td>
          <td>...</td>
          <td>T</td>
          <td>C</td>
        </tr>
        <tr>
          <td>...</td>
          <td>8</td>
          <td>108596584</td>
          <td>108596584</td>
          <td>...</td>
          <td>single base substitution</td>
          <td>...</td>
          <td>T</td>
          <td>C</td>
        </tr>
        <tr>
          <td>...</td>
          <td>15</td>
          <td>34737692</td>
          <td>34737692</td>
          <td>...</td>
          <td>single base substitution</td>
          <td>...</td>
          <td>A</td>
          <td>C</td>
        </tr>
        <tr>
          <td>...</td>
          <td>9</td>
          <td>84841627</td>
          <td>84841627</td>
          <td>...</td>
          <td>single base substitution</td>
          <td>...</td>
          <td>C</td>
          <td>G</td>
        </tr>
      </table>
    </div>
    </li>
    <br />
    <li>
      <p>
      Any tsv (tab-separated values) or csv (comma-separated values) file with the following columns: chromosome, chromosome_pos, mutated_from, mutated_to (<a href="/download/example3.tsv">download sample file</a>) <br />
      </p>
      <div class="table-responsive">
        <table class="table table-borderless example-table">
          <thead>
            <tr>
              <th scope="col">chromosome</th>
              <th scope="col">chromosome_pos</th>
              <th scope="col">mutated_from</th>
              <th scope="col">mutated_to</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>6</td>
              <td>65152140</td>
              <td>T</td>
              <td>C</td>
            </tr>
            <tr>
              <td>8</td>
              <td>108596584</td>
              <td>T</td>
              <td>C</td>
            </tr>
            <tr>
              <td>15</td>
              <td>34737692</td>
              <td>A</td>
              <td>C</td>
            </tr>
            <tr>
              <td>9</td>
              <td>84841627</td>
              <td>C</td>
              <td>G</td>
            </tr>
          </tbody>
        </table>
      </div>
    </li>
  </ul>

  <p>
    QBiC-Pred will process each input file to extract the genomic sequence contexts for all variants (please ensure that the correct version of the genome is selected on the “Make Predictions” page). The extracted sequence context will be visible to the user when the output results are displayed.
  </p>

  <p>
    The input file size is currently limited to 100MB since each file is processed in our server. Since QBiC makes prediction based on base substitution mutation, if the user has rows contain other type of mutations, they can be filtered out first to enable more rows with base substitution to be processed.  Also, for large jobs containing a large number of variants, the user can check on the status of the job by using the <span class="inline-img"><img src="/static/images/about_page/recent_btn.png" /></span> button on the upper right corner of any QBiC page. The jobs have unique identifies and they are stored on the QBiC-Pred server for 48 hours. This page can be bookmarked for later visit.
  </p>

  <p>
    Other required input fields:
  </p>
  <ul>
    <li>
      <p>
        <b>TF proteins to be tested.</b>	TF proteins to be tested. The user can select either entire families of TFs (see   <span class="gray-inline-text">Select TFs by families</span>), or individual TFs by using <span class="gray-inline-text">Select transcription factors:</span>  or by providing an input file with TF gene names (see  <span class="gray-inline-text">upload a TF file</span>). <br />
        Individual TFs, as well as TF families, are searchable by the user: <br />
      </p>
      <span class="example-select"><img class="img-fluid" src="/static/images/about_page/searchbytf.png" /></span>
      <span class="example-select"><img class="img-fluid" src="/static/images/about_page/searchbyfam.png" /></span>
      <p>
        QBiC-Pred uses the <a href="https://www.genenames.org">HGNC</a> nomenclature for gene names. A full list of TF gene names for which QBiC-Pred contains predictive models is available <a href="/download/genelist.txt">here</a>. In case the user provides a file containing gene names not recognized by QBiC-Pred, or for which we do not current have predictive models, an error message will alert the user:
      <p>
      <span class="example-alert"><img class="img-fluid" src="/static/images/about_page/tfsnotfound.png" /></span>
    </li>
    <li>
      <p>
        <b>Human genome version</b> to be used (default: hg19).
      </p>
    </li>
    <li>
      <b>Desired output</b> for the QBiC-Pred results, which can be either:
      <ul>
        <li>
          <b>p-value threshold</b> (default option). In this case QBiC-Pred will return, for each input variant, all input TFs predicted to be affected by that variant at a significance (p-value) of less than the value set of the user (default 0.0001). Predictions of both increased and decreased TF binding will be returned.
        </li>
        <li>
          <b>largest absolute z-score</b>. In this case QBiC-Pred will return, for each input variant, the input TF (or the top 2 or 3 input TFs) for which the variant has the largest effect (i.e. largest z-score in absolute value). Z-scores represent normalized binding changes, and thus they are comparable among TFs. Positive z-scores reflect increased TF binding due to the variant. Negative z-scores reflect decreased binding.
        </li>
      </ul>
    </li>
  </ul>

  <a name="section2"><h3>2. TF binding models</h3></a>
  <p>QBiC predictions are based on quantitative TF binding models trained on universal protein binding microarray (PBM) data from the following databases:</p>
  <ul>
    <li>UniPROBE: <a href="http://thebrain.bwh.harvard.edu/uniprobe/">http://thebrain.bwh.harvard.edu/uniprobe/</a></li>
    <li>CisBP: <a href="http://cisbp.ccbr.utoronto.ca/">http://cisbp.ccbr.utoronto.ca/</a></li>
    <li>PBM data generated in the Gordan laboratory (reported in <a href="https://www.cell.com/cell-systems/fulltext/S2405-4712(18)30062-0">Shen et al. Cell Systems 2018</a>, or included with the release of QBiC)</li>
  </ul>
  <p>
    We downloaded the entire database from CisBP (v1.02) consisting of 3,342 universal PBM data sets. Next, we added 245 PBM data sets from the UniPROBE database, which were not included in CisBP, and 22 PBM data sets generated in our lab, resulting in a total of 3,609 PBM data sets. Then we annotated the quality of each uPBM data based on: 1) the number of 8-mers with binding enrichment score (E-scores) exceeding 0.45; and 2) the performance of our models trained on the data, which enabled us to filter out low quality PBM data sets. After the filtering process, we utilized <a href="http://humantfs.ccbr.utoronto.ca/">The Human Transcription Factors Project</a>, as well as manual curation, to map PBM data sets to human TFs. By keeping only the best quality or the highest amino-acid identity PBM data sets for each TF, we finally arrived at a total of 666 data sets characterizing the binding of 577 human TFs. The full mapping of human TFs to available PBM data sets is available <a href="/download/mapping_gene_to_upbm_final.csv">here</a>.
  </p>
  <p>
    Our TF binding models are linear regression models trained on <a href="https://www.ncbi.nlm.nih.gov/pubmed/28691125">universal PBM data<a/>. We use the counts of 6-mers or gapped 6-mers (with a gap of 1 or 2 bases) as features, and the binding intensity signal as the outcome variable. Then, we use ordinary least squares (OLS) to estimate the coefficients for all 6-mer features, as well as the distributions of these estimates. The effect of each mutation can be written as a linear combination of all the coefficients of 6-mers covering it and can therefore be estimated as well as tested for statistical significance, as described in detail in our <a href="https://www.ncbi.nlm.nih.gov/pubmed/28691125">previous work</a>. Thus, for each variant and each PBM data set for a TF protein, our model predicts:
  </p>
  <ul>
    <li><b>Difference</b> = the difference in TF binding (i.e. in PBM signal) between the two alleles
    <li><b>z-score</b> = the normalized difference in TF binding, computed based on the estimated parameter distributions; the z-scores are comparable among different TFs and data sets</li>
    <li><b>p-value</b> = the significance of the predicted difference, computed based on the estimated parameter distributions; the p-values depend on the magnitude of difference, as well as the quality of the model and the training data.</li>
  </ul>
  <br />

  <a name="section3"><h3>3. QBiC-Pred output</h3></a>
  <p>QBiC-Pred reports results in the form of a table that the user can explore and interact with:</p>
  <span class="screenshoot-img"><img class="img-fluid" src="/static/images/about_page/output1.png" /></span>

  <p>
    In addition to reporting the TF(s) affected by each variant, QBiC provides information about the status of each allele, which can be “bound”, “unbound”, or “ambiguous”. This status is derived from the same PBM data sets used to train the OLS binding models. But, unlike the OLS model, defining the status of an allele is based on 8-mer enrichment scores (E-scores) commonly derived from PBM data (see, <a href="https://www.ncbi.nlm.nih.gov/pubmed/16998473">Berger et al. Nature Biotech, 2006</a>; <a href="https://www.cell.com/cell-systems/fulltext/S2405-4712(18)30062-0">Shen et al. Cell Systems 2018</a>). As in our previous work, we define an allele as specifically “bound” if it is contained in two overlapping 8-mers with E-score &gt; 0.4. Conversely, we define an allele as “unbound” if all 8-mers overlapping the allele have E-score &lt; 0.35. All other cases are considered “ambiguous”.
  </p>

  <p>
    QBiC provides this information to allow users to focus only on prediction that are of most interest to them, e.g. from “unbound” for the reference allele to “bound” for the alternative allele (“unbound&gt;bound” in the “Binding status” column). The users can easily filter the results by clicking on values in the “Binding status” column:
  </p>

  <span class="combined-img"><img class="img-fluid" src="/static/images/about_page/output2.png" /></span>

  <p>
    Filters can be set in a similar manner for TF names, models, and PBM data sets. For example:
  </p>

  <span class="combined-img"><img class="img-fluid" src="/static/images/about_page/output3.png" /></span>

  <p>
    For PBM data sets (column  <span class="gray-inline-text">PBM filename</span>), the QBiC output table also provides an option to directly download the data and the model used to make the predictions. The output table also specifies which type of model, i.e. using gapped or ungapped k-mers is used (column  <span class="gray-inline-text">Gap model</span>).
  </p>

  <p>
    Additional output processing options include filtering by p-value, z-score, sequence, or TF name using the “Search” functionality:
  </p>

  <div class="screenshoot-img"><img class="img-fluid" src="/static/images/about_page/output4a.png" /></div>
  <p>
    For example, if the user is interested in visualizing only predictions with large increased in TF binding, he/she can set minimum z-score value such as depicted below:
  </p>
  <div class="screenshoot-img"><img class="img-fluid" src="/static/images/about_page/output4b.png" /></div>

  <p>
  At any point during the visualization and processing of the QBiC-Pred results, users can download the current results in comma delimited format using <span class="gray-inline-text">download <font color="blue">csv</font>/<font color="blue">tsv</font></span> link below the table. The downloaded results can be uploaded back into QBiC-Pred at any time for further processing and filtering, using the “Upload Result” page:
  </p>
  <div class="screenshoot-img"><img class="img-fluid" src="/static/images/about_page/uploadpage.png" /></div>
  <p>This uploaded data will be saved for 24 hours, after which the user can upload the result file again in order to use the website interface to explore the results.</p>
  <br />
</div>
{% endblock %}
